Axom Dev Guide Updates

rhornung67 opened the pull request:
# Summary

- This PR clarifies some of the Axom Dev Guide discussion and add some details related to TPLs and updates of those.
- It also update some of the release process discussion.

The RTD content for this PR is here: https://axom.readthedocs.io/en/task-rhornung67-devguide-tpls/docs/sphinx/dev_guide/index.html

rhornung67 commented on src/docs/sphinx/dev_guide/release.rst:
```diff
@@ -39,41 +39,65 @@ Here are the steps to follow for an Axom release.
+   <https://github.com/LLNL/axom/issues/new>`_ and note it in the ``known bugs``
+   section of the release notes. Alternatively, if time permits, fix the 
+   bug in a different branch and create a pull request as you would do during
+   regular development. After the bus is resolved and that pull request is
```
Thanks. Fixed.

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -22,77 +23,82 @@ install and update these dependencies for Axom.
+and configurations of software packages. It has recipes for building each package
+with available variants to customize it to your needs. For example, Axom
+has variants for Fortran, MPI, and others. The recipes drive
+the individual packages build systems and manage packages they depend on.
```
```suggestion
the individual packages' build systems and manage packages they depend on.
```
package, plural, possessive

rhornung67 replied:
Thanks. Fixed.

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -22,77 +23,82 @@ install and update these dependencies for Axom.
+the individual packages build systems and manage packages they depend on.
 Spack also handles system level packages, so you can describe where they are on your
-system instead of building them from scratch.  You will need to describe which compilers
+system instead of building them from scratch. You will need to describe which compilers
```
What?  Only single-space between sentences? : )

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -22,77 +23,82 @@ install and update these dependencies for Axom.
+
+* Axom specific Spack package files live under ``axom/scripts/spack/packages``. These override
+  the package files in Spack's repository under ``var/spack/repos/builtin/packages``.
+  We try to minimize these, but we often have to alter the existing Spack packages to: apply fixes
+  before pushing them up to Spack proper, or alter recipes in ways that are Axom specific.
+  Such overrides do not happen at the Spack level, but at the next level, Uberenv 
+  described below.
```
I suggest:
- remove the colon from "to: apply fixes"
- remove the comma from "up to Spack proper, or alter recipes"
- either add a comma, making it "the next level, Uberenv, described below," or add parentheses, making it "the next level, Uberenv (described below)."

The last point is needed because "described below" is an adjective phrase modifying "Uberenv."  Without the comma or the parens, overrides happen at the noun phrase "Uberenv described below" which makes less sense.

rhornung67 replied:
Got it. Change made.

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -101,68 +107,76 @@ Axom only uses it for our Windows TPL builds.
+* ``build_tpls.py``: First, this script builds a set of TPLs for each of the specs
+  listed in the ``specs.json`` file for the platform on which it is run. For each TPL set,
+  it will generate a host-config file and copy it to the top-level directory of the local
+  copy of the Axom repository in which the script is run. After building all of TPL sets
+  for a platform, it will test Axom against those TPL installs as well as the ``using-with-cmake``
   example for correctness. This script stops at the first failed TPL build but
```
```suggestion
  copy of the Axom repository in which the script is run. After building all TPL sets
  for a platform, it will attempt to build Axom and the ``using-with-cmake`` example against 
  each set of TPLs. This script stops at the first failed TPL build but
```

rhornung67 replied:
Recommended change pushed.

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -101,68 +107,76 @@ Axom only uses it for our Windows TPL builds.
-   to before it redirects the output from the command, i.e., ``[[log file: /path/to/log``.
+   Due to the large amount of information printed to the screen during a full build, the build scripts
+   redirect most build step output to log files. This output will tell you what command is being run,
+   i.e., ``[EOE: some/command --with-options]``, and will tell you the log file being written
```
EOE?

rhornung67 replied:
Good lord. What was I doing?!?

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -172,105 +186,119 @@ Updating TPLs
 #. Download the new release and override the source that is already there.
-   This can often involve removing files no-longer needed but most of the
-   current ones are a single header file.
+   This may involve removing files no-longer needed.
```
```suggestion
   This may involve removing files no longer needed.
```
(minor suggestion for style and clarity)

rhornung67 replied:
Fixed.

agcapps commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -172,105 +186,119 @@ Updating TPLs
 with a new set of TPLs. Typically, this process is followed when you want to 
-update one or more TPLs which Axom depends on. After they are built and
-the required changes are merged into develop, they will be available for
+update one or more TPLs on which Axom depends. After they are built and
```
Possibly:
```suggestion
update one or more TPLs. After they are built and
```
They're our TPLs.  We don't have to say again that we depend on them.

rhornung67 replied:
Thank you. Fixed.

agcapps reviewed (approved):
Thanks, Rich!  Much clearer.

kennyweiss commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -101,68 +107,76 @@ Axom only uses it for our Windows TPL builds.
+  the end indicating which Axom build succeeded or failed.
+* ``build_src.py``: This script uses the existing host-configs in your local clone of the
+  Axom repo, or a specific one you point at, and builds and tests Axom against them. It also
+  tests the ``using-with-cmake`` example.
```
Possibly helpful clarification:

```suggestion
  tests axom installation via the ``using-with-cmake`` example and axom tutorials.
```

rhornung67 replied:
Got it. Thank you.

kennyweiss commented on src/docs/sphinx/dev_guide/updating_tpls.rst:
```diff
@@ -305,9 +333,18 @@ other Axom developers to use during development, in Axom GitLab CI testing, etc.
+   When you are confident that everything is correct and working, you will need
+   to perform this step on each of the machines named in Axom's standard host-configs.
+
+   .. important:: To install TPL builds to be shared by all Axom developers and used 
+                  in our GitLab CI, you will need to become the Axom service user ``atk``.
+                  There is a clone of the Axom repo in the ``/usr/workspace/atk/axom_repo``
+                  directory. After becoming ``atk``, you can go into that directory and
+                  switch to the branch you made to test your changes. Before running the
+                  TPL builds, make sure the branch is updated, including all submodules.
```
Should we mention that this is only applicable to the subset of axom developers who can become `atk` ?

rhornung67 replied:
I think that's implied since the /usr/workspace/atk directory is not accessible if you do not `xsu atk` and only members of the team can do that.  

The main point I was trying to make is that you really need to become atk and use that clone of the repo to build and install the shared TPLs.

kennyweiss reviewed (approved):
Thanks @rhornung67 for this thorough update to our build and release process!

rhornung67 commented:
I've addressed all the reviewer comments so far. Thank you for the corrections and suggestions. If anyone else wants to review this, please do that soon. I will merge it when all checks are green.
